llvm/lib/Transforms/Utils/SimplifyIndVar.cpp

```diff
@@ -1384,7 +1384,22 @@ WidenIV::getExtendedOperandRecurrence(WidenIV::NarrowIVDefUse DU) {
   else if (ExtKind == ExtendKind::Zero && OBO->hasNoUnsignedWrap())
     ExtendOperExpr = SE->getZeroExtendExpr(
       SE->getSCEV(DU.NarrowUse->getOperand(ExtendOperIdx)), WideType);
-  else
+  else if (DU.NeverNegative) {
```

Comment:
I would write the code here like this, to avoid duplicating the logic:
```
  if (OBO->hasNoSignedWrap() && (ExtKind == ExtendKind::Sign || DU.NeverNegative)) {
    ExtKind = ExtendKind::Sign;
    ExtendOperExpr = SE->getSignExtendExpr(
SE->getSCEV(DU.NarrowUse->getOperand(ExtendOperIdx)), WideType);
  } else if (OBO->hasNoUnsignedWrap() && (ExtKind == ExtendKind::Zero || DU.NeverNegative)) {
    ExtKind = ExtendKind::Zero;
    ExtendOperExpr = SE->getZeroExtendExpr(
SE->getSCEV(DU.NarrowUse->getOperand(ExtendOperIdx)), WideType);
  } else
    return {nullptr, ExtendKind::Unknown};
```

Reply:
Hm, I guess that wouldn't do the right thing if the operation is both nuw and nsw...

Reply:
Done post commit in 7c93452e174dd182c36471bc1e8272f26c0ae6db